#88 Add applying blocks via diffusion and comment out old mechanism

paweljakubas opened the pull request:
#88 

# Overview

- [x] I have added getBlocks and requestTip support in diffusion layer
- [x] The additional thread is instantiated that is responsible for downloading blocks from the node
- [x] The added mechanism is based on pull paradigm, ie., the wallet is the initiator for pulling blocks, makes sure the blocks are in line with what the node has
- [x] The same resolved blocks are created using the added mechanism as it is in case of old mechanism
- [x] apply blocks part of old mechanism is commented out  
- [x] I have handled a number of pathological cases, like corrupt data coming from the node or no communication with the node for some time 
- [x] Added `pull-mechanism` flag to switch on pull mechanism. Otherwise, push mechanism, the readerT based - well tested but introducing coupling, is used 

# Comments

The thorough description of the proposed mechanism are
[here](https://github.com/input-output-hk/cardano-wallet/wiki/ApplyingBlocks-Rollback-After-Decoupling-Byron) 

# Testing
## unit tests
When run as it is push mechanism is used as default. In order to run unit tests in **pull mechanism** first change in `test/unit/Test/Spec/Fixture.hs`:
```
41  usePullMechanism :: Bool                                                                                                                                                                  
42  usePullMechanism = True
```
and then 
```
$ stack test cardano-wallet:test:unit
```

## integration tests
In order to run integration tests using pull mechanism first change in `test/integration/Test/Integration/Framework/Cluster.hs` :
```
72      , ("PULL_MECHANISM", "True")
``` 
and then
```
$ stack test cardano-wallet:test:integration
```
One should see the following log in `edge.log.pub`:
```
[pawel@arch logs]$ ag "mechanism" edge.log.pub
43602:[cardano-sl.edge:Debug:ThreadId 1063] [2019-01-26 19:07:54.34 UTC] pull mechanism (diffusion layer based) for block syncing is active
```
In default case one should see instead : 
 ```
[pawel@arch logs]$ ag "mechanism" edge.log.pub
538:[cardano-sl.edge:Debug:ThreadId 1163] [2019-01-26 18:57:03.74 UTC] push mechanism (readerT based) for block syncing is active
```

## wallet-node syncing with mainnet
When running wallet with node attached to mainnet see [here](https://github.com/input-output-hk/cardano-wallet/wiki/Building-Running-and-Testing-a-wallet-node)
In order to switch on **pull mechanism** of block syncing use additionally `--pull-mechanism```
One should see the same logs as above confirming switching on the pull mechanism


A number of problems were solved during this PR:
**Update1**
Now after the metaTx fix coin selection tests in Transactions testsuite and Utxo test in Wallets testsuite are alright. Occasionally, I am witnessing the thread responsible for block application to be stuck (and then all tests that rely on block application fail) for the reason I do not understand at this moment. 

**Update 2**
I addressed those rare cases : it was due to (a) no communication with node, (b) corrupt data flowing from the node, (c) combination of (a) and (b). 
I tested using integration tests more than 100 - found all cases and handled them properly (ie., integration tests pass)
 
Very occasionally the node response of `requestTip` (from `lib/src/Pos/Diffusion/Full/Block.hs` from `cardano-sl`): 
```
479	    handleTip _ t = do
480	        traceWith logTrace (Warning, sformat ("requestTip: got enexpected response: "%shown) t)
481	        throwIO $ DialogUnexpected "peer sent more than one tip"
```

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas commented on src/Cardano/Wallet/WalletLayer/Kernel.hs:
```diff
@@ -82,12 +103,26 @@ bracketPassiveWallet pm mode logFunction keystore node fInjects f = do
-                 { Actions.applyBlocks = \blunds -> do
+                 --{ Actions.applyBlocks = \blunds -> do
+                { Actions.applyBlocks = \_ -> do
+                    return ()
```
Commenting out readerT mechanism

paweljakubas commented on src/Cardano/Wallet/Server/Plugins.hs:
```diff
@@ -64,6 +66,10 @@ import           Pos.Util.Wlog (logInfo, modifyLoggerName, usingLoggerName)
+
+instance Given SscConfiguration where
+    given = sscConfiguration
+
```
need this in order to use getBlocks and service MonadBlockVerify constraint

KtorZ replied:
We will find a way around this :) ... Having it passed as an extra parameter isn't a bad choice. The reflection stuff is... not pragmatic.

paweljakubas replied:
I will leave it for another PR

paweljakubas commented on src/Cardano/Wallet/Kernel/Submission/Worker.hs:
```diff
@@ -43,3 +46,29 @@ tickSubmissionLayer logFunction tick =
+                      -- ^ A function to call at each 'tick' of the worker.
+                      -- This callback will be responsible for doing any pre
+                      -- and post processing of the state.
+                   -> (([HeaderHash],[HeaderHash]),Utxo)
```
This will probably be different types as we will switch to checkpoint based mechanism

paweljakubas commented on src/Cardano/Wallet/Kernel/DB/BlockContext.hs:
```diff
@@ -46,7 +46,7 @@ data BlockContext = BlockContext {
       -- See 'mostRecentMainBlock'.
     , _bcPrevMain :: !(Strict.Maybe (InDb Core.HeaderHash))
-    } deriving Eq
+    } deriving (Eq, Show)
```
needed this to print resolved blocks - will clean this in the final version

paweljakubas commented on src/Cardano/Wallet/Kernel/DB/Resolved.hs:
```diff
@@ -105,7 +105,7 @@ data ResolvedBlock = ResolvedBlock {
       -- | Creation time of this block
     , _rbMeta    :: !Timestamp
-    }
+    } deriving Show
```
needed this to print resolved blocks - will clean this in the final version

paweljakubas commented on src/Cardano/Wallet/WalletLayer/Kernel.hs:
```diff
@@ -165,9 +200,23 @@ bracketActiveWallet
     Kernel.bracketActiveWallet passiveWallet walletDiffusion $ \w -> do
+
+        genesisConfig <- liftIO $ getCoreConfig $ passiveWallet ^. Kernel.walletNode
+        let initialUtxo = ccUtxo $ initCardanoContext genesisConfig
```
I have to do it in order to have initial Utxo - in the readerT mechanism it comes with Undos

paweljakubas commented on src/Cardano/Wallet/WalletLayer/Kernel.hs:
```diff
@@ -82,12 +103,26 @@ bracketPassiveWallet pm mode logFunction keystore node fInjects f = do
+                                       ) ls--}
+                    let mp = Debug.trace (("####### WalletActionInterp:\n" <> show ls)::String) $ catMaybes ls
+                    --let mp = catMaybes ls
                     mapM_ (Kernel.applyBlock w) mp
```
here it was the place where BListener applyBlocks was invoked

paweljakubas commented on src/Cardano/Wallet/WalletLayer/Kernel.hs:
```diff
@@ -165,9 +200,23 @@ bracketActiveWallet
+            (passiveWallet ^. Kernel.walletLogMessage)
+            tickDiffusionFunction
+            (([],[]), initialUtxo)
+
```
Instantiating dedicated diffusion ticker thread

KtorZ commented on src/Cardano/Wallet/Kernel/Diffusion.hs:
```diff
@@ -33,13 +40,76 @@ data WalletDiffusion = WalletDiffusion {
+            in
+            \nodeId from toNotIncluding -> do
+                 case toNotIncluding of
+                     [lastConsumedHeader] -> do
```
Why is this a list if we only care about this being a singleton or.. something else? Wouldn't a `Maybe` better reflect the intent here?

paweljakubas replied:
If servicing applying blocks only using `Maybe` would be the best. But, I would like to reuse this call in case of rollback, and in that case I will deliver here **N lastly downloaded headers*

KtorZ commented on src/Cardano/Wallet/Kernel/Submission/Worker.hs:
```diff
@@ -43,3 +46,29 @@ tickSubmissionLayer logFunction tick =
+
+
+tickDiffusionLayer :: forall m. (MonadCatch m, MonadIO m)
+                   => (Severity -> Text -> m ())
```
Please, no more variable-length indentation :(

paweljakubas replied:
done

KtorZ commented on src/Cardano/Wallet/WalletLayer/Kernel.hs:
```diff
@@ -180,3 +229,90 @@ bracketActiveWallet walletPassiveLayer passiveWallet walletDiffusion runActiveLa
+    tickDiffusionFunction
+        :: (([HeaderHash], [HeaderHash]), Utxo)
+        -> IO (([HeaderHash], [HeaderHash]), Utxo)
+    tickDiffusionFunction ((inProgressHeader, consumedHeader),utxo) = do
```
This sounds like something that should be in `Kernel/Diffusion` instead.

paweljakubas replied:
done

KtorZ commented on test/integration/Test/Integration/Scenario/Transactions.hs:
```diff
@@ -7,9 +7,31 @@ import           Universum
+            noSpendingPassword
+
+        --need this to see transition between epochs
+        liftIO $ threadDelay $ 120 * oneSecond
```
I suggest introducing a dedicated DSL extension for this `expectNextEpochEventually` or something like this.

paweljakubas replied:
I will do it in the separate PR

KtorZ commented on src/Cardano/Wallet/WalletLayer/Kernel.hs:
```diff
@@ -180,3 +229,90 @@ bracketActiveWallet walletPassiveLayer passiveWallet walletDiffusion runActiveLa
+                let utxoAfterAddition = utxoUnions $ utxo : utxoToAdd
+                let utxoSetToDel = Set.fromList $ concat $ map fst $ catMaybes txPayload
+                let updatedUtxo = utxoRemoveInputs utxoAfterAddition utxoSetToDel
+                pure $ (([h1], (reverse accomodatedHeaders) ++ consumedHeader), updatedUtxo)
```
Wow. That's one big function which I honestly didn't try to read much :( ... Sounds like this would deserve to be clarified and decomposed a bit in smaller pieces.

paweljakubas replied:
simplified

paweljakubas commented on src/Cardano/Wallet/Kernel/Diffusion.hs:
```diff
@@ -1,3 +1,4 @@
+{-# OPTIONS_GHC -fno-warn-type-defaults #-}
```
yes, I need this for temporary logging

paweljakubas commented on src/Cardano/Wallet/Kernel/Diffusion.hs:
```diff
@@ -33,13 +40,76 @@ data WalletDiffusion = WalletDiffusion {
+                        _ -> do
+                            -- unexpected so returning current blocks in the batch
+                            --void $ print "N"
+                            pure $ OldestFirst blocksInThisBatch --
```
As underlined above I use `getPrevBlock nodeId from toMaybe blocksInThisBatch` in a recursive fashion. Inside diffusion layer's `getBlocks` with `from` is used. This line is "insurance" that trusted node gives `[]` for the reason I cannot really anticipate here

paweljakubas commented on src/Cardano/Wallet/Kernel/Diffusion.hs:
```diff
@@ -33,13 +40,76 @@ data WalletDiffusion = WalletDiffusion {
+                    downloadedBlock <- getBlocks d nodeId from [from]
+                    let downloadedBlockStripped = getOldestFirst downloadedBlock
+                    --void $ print toMaybe
+                    case downloadedBlockStripped of
```
done

paweljakubas commented on src/Cardano/Wallet/Kernel/Diffusion.hs:
```diff
@@ -33,13 +40,76 @@ data WalletDiffusion = WalletDiffusion {
+                            --void $ print "N"
+                            pure $ OldestFirst blocksInThisBatch --
+            in
+            \nodeId from toNotIncluding -> do
```
done

paweljakubas commented:
after successful addressing review remarks I will rebase branch against develop and squash the commits

paweljakubas commented:
rebased

paweljakubas commented:
rebased and added flag pull-mechanism to switch on this mechanism. Otherwise it readerT mechanism, push one, well tested but introducing coupling, is used

paweljakubas commented:
When ok, I will squash the commit

Anviking reviewed (approved):
The `Typeable` stuff in the tests looks a bit awkward, but I _think_ it makes sense still 👍
